staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/mutating/reinvocationcontext.go

```diff
@@ -0,0 +1,76 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mutating
+
+import (
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+type key struct {
+	PolicyUID     types.NamespacedName
+	BindingUID    types.NamespacedName
+	ParamUID      types.NamespacedName
+	MutationIndex int
+}
+
+type policyReinvokeContext struct {
+	// lastPolicyOutput holds the result of the last Policy admission plugin call
+	lastPolicyOutput runtime.Object
+	// previouslyInvokedReinvocablePolicys holds the set of policies that have been invoked and
+	// should be reinvoked if a later mutation occurs
+	previouslyInvokedReinvocablePolicies sets.Set[key]
+	// reinvokePolicies holds the set of Policies that should be reinvoked
+	reinvokePolicies sets.Set[key]
+}
+
+func (rc *policyReinvokeContext) ShouldReinvoke(policy key) bool {
+	return rc.reinvokePolicies.Has(policy)
+}
+
+func (rc *policyReinvokeContext) IsOutputChangedSinceLastPolicyInvocation(object runtime.Object) bool {
+	return !apiequality.Semantic.DeepEqual(rc.lastPolicyOutput, object)
+}
+
+func (rc *policyReinvokeContext) SetLastPolicyInvocationOutput(object runtime.Object) {
+	if object == nil {
+		rc.lastPolicyOutput = nil
+		return
+	}
+	rc.lastPolicyOutput = object.DeepCopyObject()
+}
+
+func (rc *policyReinvokeContext) AddReinvocablePolicyToPreviouslyInvoked(policy key) {
```

Comment:
Seeing `// if preconditions are not met, then skip mutations` in the dispatcher, skipping the call of this method. I.e. one policy mutating in a way that another policy applies does not cause the later to be called. Is that intentional?

Reply:
> what's the semantics here? If a policy is not applied due to some match conditions, but for an reinvocation a match conditions would actually match, would that policy show up here or get ignored?

No, this is not supported for match conditions (for webhooks or for policies).  I had not realize this is a gap until now! I've added a testing showing the gap.  I'd be happy to look into it.  OK if we follow up post-freeze, make this beta graduation criteria?

Reply:
> one policy mutating in a way that another policy applies does not cause the later to be called. Is that intentional?

I've also added a test showing that if the first policy adds a label that the next policy has a match condition for, that the second policy is evaluated.  But my above comment applies to the general case.

Reply:
Making it a beta graduation criteria sounds fine. Am not even sure it is that clear which behaviour we want. Reinvocation is always kind of best effort. Sometimes it might need more than one reinvocation to converge, which we obviously never do. So deciding and documenting is also an option for the behaviour here.